hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala

```diff
@@ -248,5 +272,105 @@ abstract class PBaseStruct extends PType {
     }
   }
 
+  def deepCopyFromAddress(mb: MethodBuilder, region: Code[Region], srcStructAddress: Code[Long]): Code[Long] = {
+    val dstAddress = mb.newField[Long]
+    Code(
+      dstAddress := this.copyFrom(mb, region, srcStructAddress),
+      this.fastDeepCopyLoop(mb, region, dstAddress),
+      dstAddress
+    )
+  }
+
+  def fastDeepCopyLoop(mb: MethodBuilder, region: Code[Region], dstStructAddress: Code[Long]): Code[Unit] = {
+    var loop: Code[Unit] = Code._empty
```

Comment:
wouldn't call this loop, since it's not generating a loop.

Reply:
I'm a bit confused by what you mean. It is, the body of the function is just a loop over elements/fields, potentially with recursion (hence deep)

Reply:
I think you mean that the emitted code is unrolled, that’s true but loop unrolling feels like an implementation detail (for instance the interpreted code will definitely emit a loop, because the compiler doesn’t statically know size, although  the JIT may figure it out at runtime and unroll it)

Reply:
Happy to change the name however